wi-fi/hal/cyhal_sdio.c

```diff
@@ -4,766 +4,128 @@
+	if (command != CYHAL_SDIO_CMD_IO_RW_DIRECT) {
 		*response = 0;
-
-	val = *(sdio_common.base + pres_state);
-	if (val & 0x7) {
-		cy_log_msg(CYLF_SDIO, CY_LOG_ERR, "cannot issue SDIO bulk transfer cmd (pres_state=%x)\n", val);
-		return CYHAL_SDIO_RSLT_ERR_BAD_PARAM;
+		return 0;
 	}
 
-	val = *(sdio_common.base + int_status);
-	if (val & 0x2) {
-		cy_log_msg(CYLF_SDIO, CY_LOG_ERR, "cannot issue SDIO bulk transfer cmd (int_status=%x)\n", val);
-		return CYHAL_SDIO_RSLT_ERR_BAD_PARAM;
-	}
-
-	if (length > DMA_BUFFER_SIZE) {
-		cy_log_msg(CYLF_SDIO, CY_LOG_ERR, "cannot issue SDIO bulk transfer cmd (length=%u > %u)\n", length, DMA_BUFFER_SIZE);
-		return CYHAL_SDIO_RSLT_ERR_BAD_PARAM;
-	}
-
-	if (direction == CYHAL_WRITE)
-		memcpy(sdio_common.dmaptr, data, length);
-
-	cmd = CYHAL_SDIO_CMD_IO_RW_EXTENDED << 24;
-	cmd |= 1 << 21;   /* DPSEL=1 */
-	cmd |= 1 << 20;   /* CICEN=1 */
-	cmd |= 1 << 19;   /* CCCEN=1 */
-	cmd |= 0x2 << 16; /* RSPTYP=2 */
-
-	mix = (1U << 31) | (1 << 0); /* DMAEN=1 */
-	if (direction == CYHAL_READ)
-		mix |= 1 << 4; /* DTDSEL=1 */
-
-	if (block_mode) {
-		mix |= (1 << 5); /* MSBSEL=1 */
-		mix |= (1 << 1); /* BCEN=1 */
-
-		blk = (count << 16) | 64; /* BLKCNT=count BLKSIZE=64 */
-	}
-	else {
-		blk = (1 << 16) | count; /* BLKCNT=1 BLKSIZE=count */
-	}
-
-	*(sdio_common.base + mix_ctrl) = mix;
-	*(sdio_common.base + blk_att) = blk;
-	*(sdio_common.base + ds_addr) = sdio_common.dmaphys;
-	*(sdio_common.base + cmd_arg) = argument;
-	*(sdio_common.base + cmd_xfer_typ) = cmd;
-
-	/* wait 1 ms max */
-	if (wait_for_cmd(0xb, 10, 100) < 0) {
-		reset_cmd_block();
-		reset_data_block();
-		return CYHAL_SDIO_RSLT_ERR_FUNC_RET(CYHAL_SDIO_RET_CMD_TIMEOUT);
-	}
+	data = arg.cmd52.data;
+	rslt = sdio_transferDirect(arg.cmd52.dir, arg.cmd52.address, arg.cmd52.areaIndex, &data);
+	*response = data;
 
-	/* clear status flags */
-	*(sdio_common.base + int_status) = 0xb; /* DINT=1 TC=1 CC=1 */
-
-	if (direction == CYHAL_READ)
-		memcpy(data, sdio_common.dmaptr, length);
-
-	/* retrieve response */
-	val = *(sdio_common.base + cmd_rsp0);
-
-	if (response)
-		*response = val;
-
-	return CY_RSLT_SUCCESS;
+	return rslt;
 }
 
-/* NOTE: obj is ignored - state is kept in sdio_common */
+
+/* NOTE: obj is ignored, state instance is held by the SDIO API*/
 cy_rslt_t cyhal_sdio_bulk_transfer(cyhal_sdio_t *obj, cyhal_transfer_t direction, uint32_t argument,
 	uint32_t *data, uint16_t length, uint32_t *response)
 {
-	cy_rslt_t res;
-
-	// cy_log_msg(CYLF_SDIO, CY_LOG_DEBUG, "cyhal_sdio_bulk_transfer (dir=%u arg=%u length=%u)\n", direction, argument, length);
-
-	for (unsigned int i = 1; i <= 10; ++i) {
-		mutexLock(sdio_common.cmd_lock);
-		res = cyhal_sdio_bulk_transfer_internal(direction, argument, data, length, response);
-		mutexUnlock(sdio_common.cmd_lock);
-
-		if (res == CY_RSLT_SUCCESS)
-			break;
-
-		cy_log_msg(CYLF_SDIO, CY_LOG_ERR, "repeating SDIO bulk transfer cmd (cnt=%d)\n", i);
-
-		usleep(1000);
-	}
-
+	size_t len;
+	sdio_cmd_arg_t arg;
```

Comment:
It could be done in an even smarter way:
```suggestion
	whd_bus_sdio_cmd53_argument_t arg = (whd_bus_sdio_cmd53_argument_t)argument;

```